firebase-firestore/src/main/java/com/google/firebase/firestore/UserDataWriter.java

```diff
@@ -0,0 +1,133 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore;
+
+import androidx.annotation.Nullable;
+import androidx.annotation.RestrictTo;
+import com.google.firebase.Timestamp;
+import com.google.firebase.firestore.model.DatabaseId;
+import com.google.firebase.firestore.model.DocumentKey;
+import com.google.firebase.firestore.model.value.ArrayValue;
+import com.google.firebase.firestore.model.value.FieldValue;
+import com.google.firebase.firestore.model.value.ObjectValue;
+import com.google.firebase.firestore.model.value.ReferenceValue;
+import com.google.firebase.firestore.model.value.ServerTimestampValue;
+import com.google.firebase.firestore.model.value.TimestampValue;
+import com.google.firebase.firestore.util.Logger;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Converts Firestore's internal types to the Java API types that we expose to the user.
+ *
+ * @hide
+ */
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+public class UserDataWriter {
+  private final FirebaseFirestore firestore;
+
+  /** Holds settings that define field value deserialization options. */
+  static class UserDataWriterOptions {
```

Comment:
I have no idea if this is a good idea, but what do you think of making these things properties of the UserDataWriter itself?

We already have to allocate a `UserDataWriterOptions` object so it's not like we're avoiding heap allocations with the current scheme. Meanwhile it would save us having to pass the options up and down every call.

This would allow you to remove the `getUserDataWriter` method on `FirebaseFirestore` and the associated mocking weirdness.

What do you think?

Reply:
The version of the PR that I showed you yesterday (that addressed both UserDataConversion and CanonicalIds) actually did just that. It was needed because I did not want to expose `options` in the canonical ID API. I scrapped this since I didn't want to create the extra object. You are correct though that I am still creating an extra object - and we will likely not be able to get rid of it since `timestampsInSnapshots` will have to stay around even when we drop in from Settings, as `DocumentSnapshot.getDate()` relies on it.

I reverted the parts of the last edit that removed the constructor params for `serverTimestampBehavior` and `timestampsInSnapshotsEnabled`.